crates/sui-framework/sources/LockedCoin.move

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module Sui::LockedCoin {
+    use Sui::Balance::{Self, Balance};
+    use Sui::Coin::{Self, Coin};
+    use Std::Errors;
+    use Sui::ID::{Self, VersionedID};
+    use Sui::Transfer;
+    use Sui::TxContext::{Self, TxContext};
+
+    /// The locked_until time passed into the creation of a locked coin is invalid.
+    const EINVALID_LOCK_UNTIL: u64 = 0;
+    /// Attempt is made to unlock a locked coin that has not been unlocked yet.
+    const ECOIN_STILL_LOCKED: u64 = 1;
+
+    /// A coin of type `T` locked until `locked_until_epoch`.
+    struct LockedCoin<phantom T> has key, store {
+        id: VersionedID,
+        balance: Balance<T>,
+        locked_until_epoch: u64
+    }
+
+    /// Returns the epoch until which the coin is locked.
+    public fun locked_until_epoch<T>(locked_coin: &LockedCoin<T>) : u64 {
+        locked_coin.locked_until_epoch
+    }
+
+    /// Wrap a balance into a LockedCoin.
+    public fun from_balance<T>(balance: Balance<T>, locked_until_epoch: u64, ctx: &mut TxContext): LockedCoin<T> {
+        LockedCoin { id: TxContext::new_id(ctx), balance, locked_until_epoch }
+    }
+
+    /// Destruct a LockedCoin wrapper and keep the balance.
+    public fun into_balance<T>(coin: LockedCoin<T>): Balance<T> {
```

Comment:
> This also somewhat unlocks the coin, right?

Good point. I added this function here because `Delegation` object contains a `Balance<SUI>`. I wanted to use the same `Delegation` struct for delegating both `Coin<SUI>` and  `LockedCoin<SUI>`. So I needed to convert a LockedCoin<SUI> to Balance<SUI> in `Delegation::create_from_locked_coin`. 

But we don't want modules other than the governance contracts to be able to call this `into_balance` function with `SUI` type. One way to do this that I can think of is to add a witness of type `&T` to this function so the signature becomes `public fun into_balance<T>(coin: LockedCoin<T>, _witness: &T): Balance<T>` and make `SUI` type accessible to the governance contracts so that only these contracts can call `LockedCoin::into_balance<SUI>`. Alternatively we can just have a separate `DelegationWithLockedCoin` struct that contains a locked coin instead of balance. What do you think?

Reply:
That might over-complicate things. To think from API-level, what you really want to expose are two functions:
- unlock
- unlock_and_take
where `unlock` just unlocks the coin, and return the balance; `unlock_and_take` unlocks and coin and send the coin to the sender.
Now with `unlock`, in delegation logic, you can simply call `unlock` to get the balance?

Reply:
> Now with unlock, in delegation logic, you can simply call unlock to get the balance?

But the issue is we want the locked SUI holder to be able to delegate their locked SUI tokens before the tokens unlock (i.e., before the `locked_until_epoch`).